### DISCUSSION TITLE
[RFC] Control watch observation depth

### DISCUSSION THREAD

**Discussion by Alfred-Skyblue** [vuejs/rfcs#597] [RFC Discussions]
Start Date: 2023-11-08
Target Major Version: 3.x
Implementation PR: https://github.com/vuejs/core/pull/9572

## Summary

Support passing watch deep option as a number type, allowing control over the depth of the watch. This helps prevent unnecessary monitoring, especially when dealing with large arrays or objects.

## Basic example

- Object

```typescript
const state = reactive({
  a: {
    b: {
      c: {
        d: {
          e: 1
        }
      }
    }
  }
})

watch(
  state,
  () => {
    console.log('state changed')
  },
  { deep: 2, flush: 'sync' }
)

state.a.b = { c: { d: { e: 2 } } } // state changed

state.a.b.c = { d: { e: 3 } } // no console

state.a.b = { c: { d: { e: 4 } } } // state changed
```

- Array

```ts
const array = ref([])
watch(
  array,
  () => {
    console.log('array changed')
  },
  { deep: 1 }
)

array.value[0] = { foo: 'foo' } // array changed
array.value[0].foo = 'bar' // watch will not be triggered

```

- watching reactive with deep: false

```ts
const state = reactive({
      foo: {
        a: 2,
      },
      bar: 'bar',
    })

    watch(
      state,
      () => {
        console.log('state changed')
      },
      { deep: false },
    )

    state.foo.a++ // watch will not be triggered

    state.bar = 'bar2' // state changed
```
[Result playground](https://deploy-preview-9572--vue-sfc-playground.netlify.app/#eNp9U8lu2zAQ/ZUBL5FrQV5yU2SjC3JogaZF2luYA01RlhKJJLg4LgT9e4dUrLTZeOFw5vHNmxmyJ5+0zg5ekJwUlptGO7DCeb2lsum0Mg56MKJK4YE5XsMAlVEdnOGNMyqp5EpaByVzDDZwcxtclTKQtMJBg67lBW4FrJa40JzPZ9BTCfFGpr2tk3gEYPkYCGuXQzPaQ9iG2QWVaJ2yMWPYH+RGWUngmYVQlJfEUArJDDbbkY8rL112YK0X8znSpFhPKYTOwRkvUrS1q3NYY5anDPHSY4Zl9FdectcoCXvhrpksVXflu50wmCmmMdgzI+E7c3VWtUqZJJomQhH0IbQAmWIZE1klVyeCqHzUebO8zUI7+9CIl/mQ4jnL+jWWFbJkO+R5QREJisU4bhw0HpzodMucwBNAsfPOIfFH3jb8fkMJ6qRky2sm92JVLMbwW9D1BF3/Dy2bw7bvH5s7DMUiOIKQKTlJibM4gqrZZ3dWSXyUsSxKuOp00wrzQ4eSLSXTc6GEta16+BZ9caYnP68Fv3/Ff2ePwUfJTyOsMAdByRRzzGC7xvDlrytxRHsKdqr0LaLfCV4Lq1ofNI6wz16WKPsfXFT7NX6tRu5/28ujE9KeigpC45uPeErwn315p/QnuefZebyHsyXDXyjgRYo=)

## Motivation

When using the `watch` function to observe arrays or objects, enabling the `deep` option is necessary for monitoring deeper levels of the object. However, this approach can lead to performance issues since it requires recursive access to all properties of the object. In such cases, the `depth` option can be employed to control the monitoring depth, thereby avoiding unnecessary observations.

## Unresolved questions

Welcome to discussion

**Comment by skirtles-code** [comment]
This is a feature idea I've pondered previously.

I think the core idea of supporting depth is a good one. It'd be particularly useful for watching changes to an array without watching objects inside the array.

Initially, I also thought that having `deep: 1`, etc. would be a good way to implement it. But on reflection I think it might be better to have a separate `depth: 1` setting.

For example:

```js
watch(..., {
  depth: 1
})
```

The above would be equivalent to:

```js
watch(..., {
  deep: true,
  depth: 1
})
```

If the user explicitly passes `deep: false`, that would take priority over the `depth`:

```js
watch(..., {
  deep: false,
  depth: 1     // The depth will be ignored as deep is false
})
```

I'm unclear whether `depth: 0` should be allowed and what it should mean. Initially, I thought that it would be equivalent to `deep: false`, but there's some subtlety around equality checking. A watcher with `deep: true` doesn't perform equality checking on the old and new values. It's not immediately clear to me whether `depth: 0` should be consistent with `deep: false` and perform the equality check, or consistent with other depths and skip the equality check. I can see arguments for either approach.

There are a few reasons why I think a separate setting might be better:

1. It'd be a bit jarring from an English perspective to have `deep: 1`.
2. The existing implementation uses a truthy/falsey check for `deep`. While TS users do get forced to use an explicit boolean, JS users don't. There's potential to introduce subtle bugs into codebases when they upgrade, if they are (perhaps inadventently) already passing a number to `deep`.
3. If we just use `deep` for this, I think it'd be too easy for someone to accidentally pass a number for `deep` when they wanted to pass `true`. There'd be no hint of a problem from the types and the code would appear to work, up to a certain depth, and then would mysteriously stop working beyond that depth. Having a separate setting makes it more difficult to make a mistake.

For points 2 and 3 above, I'm not imagining somebody passing a literal number like `deep: 2`, that'd be obvious. I'm imagining something like `deep: foo`, where it might not be so obvious that `foo` is a number.

**Reply by Alfred-Skyblue** [reply]
I agree with you. In previous versions, the user may pass `deep` as 1, but it will be considered true. I initially considered adding a property `depth`, but this will undoubtedly increase the logical complexity of watch options.

**Reply by hubdidub** [reply]
Maybe calling the additional option `maxDepth` instead of `depth` could help a little, because it makes it clear that it's not determining whether it's a deep watcher or not, it just sets a limit. Which would also mean it has no effect unless `deep: true` is also set.

**Reply by hubdidub** [reply]
Alternatively, how about this:
```js
watch(..., {
  deep: { limit: 1 } // go up to 1 level deep
})
watch(..., {
  deep: { limit: 0 } // NOT the same as deep: false, but would catch array mutations when watching an array-ref
})
watch(..., {
  deep: true // limitless, like before
})
```
Not sure if I like it because it's a nested option, but I thought I'd bring it up because I think that would also solve the complexity issue and isn't likely to break existing code.

**Reply by Alfred-Skyblue** [reply]
Thank you for the discussion. A new option, `depth`, has been added to control the depth of monitored objects.

**Reply by Alfred-Skyblue** [reply]
Temporarily decided to support passing `deep` as a number type to control the depth of the listening object. In response to the incorrect usage of `deep` as a number type in previous versions, a warning message has been added in version 3.4.

https://github.com/vuejs/core/pull/9572#issuecomment-1872529021

**Comment by Alfred-Skyblue** [comment]
After discussion, a new option, `depth`, has been introduced to control the depth of monitored objects.

**Comment by hubdidub** [comment]
I'm looking forward to this.
Controlling the observation depth for objects is really nice!

But my use case no.1 for this would be to watch for array mutations. Because we have to use deep watchers for that, and it always felt wrong to watch the entire thing when the elements are very complex objects.
Will we be able to watch the array itself but not the elements with this change?

I imagine it would be watch-level `0`, like so:
```js
const array = ref([]);
watch(array, callback, { deep: true, depth: 0 });
```

Will this trigger the callback when I do the following operations?
```js
array.value.push({ foo: 'test' });
array.value[0] = { foo: 'test2' };
array.value.splice(0, 1);
```

And this:
```js
array.value[0].foo = 'test3';
```
will _not_ trigger it, unless I set it to `depth: 1` or higher, right?

**Reply by Alfred-Skyblue** [reply]
When `depth = 0`, it is ineffective, equivalent to recursing through all properties. If you need to listen to arrays, use the following:

```ts
const array = ref([])
watch(
  array,
  () => {
    console.log('array changed')
  },
  { deep: true, depth: 1 }
)

array.value[0] = { foo: 'foo' } // array changed
array.value[0].foo = 'bar' // watch will not be triggered
```

or

```ts
const array = ref([])
watch(
  array,
  () => {
    console.log('array changed')
  },
  { deep: true, depth: 2 }
)

array.value[0] = { foo: 'foo' } // array changed
array.value[0].foo = 'bar' // array changed
```
